dmoj/cptbox/helper.cpp

```diff
@@ -117,6 +121,55 @@ int cptbox_child_run(const struct child_config *config) {
     }
 #endif
 
+#if PTBOX_LANDLOCK
+    int ruleset_fd;
+    struct landlock_ruleset_attr ruleset_attr = {
+        .handled_access_fs =
+            LANDLOCK_ACCESS_FS_EXECUTE |
+            LANDLOCK_ACCESS_FS_WRITE_FILE |
+            LANDLOCK_ACCESS_FS_READ_FILE |
+            LANDLOCK_ACCESS_FS_READ_DIR |
+            LANDLOCK_ACCESS_FS_REMOVE_DIR |
+            LANDLOCK_ACCESS_FS_REMOVE_FILE |
+            LANDLOCK_ACCESS_FS_MAKE_CHAR |
+            LANDLOCK_ACCESS_FS_MAKE_DIR |
+            LANDLOCK_ACCESS_FS_MAKE_REG |
+            LANDLOCK_ACCESS_FS_MAKE_SOCK |
+            LANDLOCK_ACCESS_FS_MAKE_FIFO |
+            LANDLOCK_ACCESS_FS_MAKE_BLOCK |
+            LANDLOCK_ACCESS_FS_MAKE_SYM,
+    };
+
+    ruleset_fd = landlock_create_ruleset(&ruleset_attr, sizeof(ruleset_attr), 0);
+    if (ruleset_fd < 0) {
+        perror("Failed to create a ruleset");
+        return PTBOX_SPAWN_FAIL_LANDLOCK;
+    }
+
+#define READ_EXACT_FILE_RULE      LANDLOCK_ACCESS_FS_EXECUTE | LANDLOCK_ACCESS_FS_READ_FILE
+#define READ_EXACT_DIR_RULE       LANDLOCK_ACCESS_FS_READ_DIR
+#define READ_RECURSIVE_DIR_RULE   READ_EXACT_FILE_RULE | READ_EXACT_DIR_RULE
+#define WRITE_EXACT_FILE_RULE     LANDLOCK_ACCESS_FS_WRITE_FILE
+#define WRITE_EXACT_DIR_RULE      LANDLOCK_ACCESS_FS_READ_DIR
+#define WRITE_RECURSIVE_DIR_RULE  WRITE_EXACT_FILE_RULE | WRITE_EXACT_DIR_RULE
+
+    if(
+        add_rules(ruleset_fd, config->read_exact_files,     READ_EXACT_FILE_RULE) ||
+        add_rules(ruleset_fd, config->read_exact_dirs,      READ_EXACT_DIR_RULE) ||
+        add_rules(ruleset_fd, config->read_recursive_dirs,  READ_RECURSIVE_DIR_RULE) ||
+        add_rules(ruleset_fd, config->write_exact_files,    WRITE_EXACT_FILE_RULE) ||
+        add_rules(ruleset_fd, config->write_exact_dirs,     WRITE_EXACT_DIR_RULE) ||
+        add_rules(ruleset_fd, config->write_recursive_dirs, WRITE_RECURSIVE_DIR_RULE)
+    ) {
+        return PTBOX_SPAWN_FAIL_LANDLOCK;
+    }
+
+    if(landlock_restrict_self(ruleset_fd, 0)) {
+    		perror("Failed to enforce ruleset");
```

Comment:
Nit: inconsistent tabs and spaces, here and in a lot of other places. We use 4-space tabs.